### PR TITLE
Fix: Jamba batched generation

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
Basically a continuation of #32677 which implements the fixes for Jamba this time. Batched generation tests might need to be changed, especially the logits, but not sure how to proceed there as the logits are HW dependent.  

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@molbap @ArthurZucker

### CODE REVIEW

File: src/transformers/models/jamba/modeling_jamba.py [hunk 1]
```diff
@@ -1383,6 +1410,17 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
...
+    def _update_mamba_mask(self, attention_mask, cache_position):
+        mamba_mask = attention_mask
+
+        # No need for zeroing states when
+        # 1. Cached forward
+        # 2. Attending to all inputs
+        if cache_position[0] > 0 or (attention_mask is not None and torch.all(attention_mask == 1)):
+            mamba_mask = None
+
+        return mamba_mask
```
Comment (vasqu) [inline src/transformers/models/jamba/modeling_jamba.py:1422]: Made this a separate function to make it more clean. Can be moved to the forward call directly too tho.

File: tests/models/jamba/test_modeling_jamba.py [hunk 2]
```diff
@@ -458,51 +458,6 @@ def test_attention_outputs(self):
...
-            pad_size = (input_ids.shape[0], 32)
-            padding = torch.ones(pad_size, dtype=input_ids.dtype, device=torch_device) * config.pad_token_id
-            padded_input_ids = torch.cat((padding, input_ids), dim=1)
-            padded_attention_mask = torch.cat((torch.zeros_like(padding), attention_mask), dim=1)
-            model_kwargs = _prepare_model_kwargs(padded_input_ids, padded_attention_mask, signature)
-            next_logits_with_padding = model(**model_kwargs).logits[:, -1, :]
-
-            # They should result in very similar logits
-            self.assertTrue(torch.allclose(next_logits_wo_padding, next_logits_with_padding, atol=3e-3))
-
```
Comment (vasqu) [inline tests/models/jamba/test_modeling_jamba.py:505]: Passed locally without the higher rtol/atol. Will see if the CI agrees.

File: tests/models/jamba/test_modeling_jamba.py [hunk 3]
```diff
@@ -707,6 +662,7 @@ def test_simple_batched_generate_with_padding(self):
         with torch.no_grad():
             logits = self.model(input_ids=inputs["input_ids"]).logits
 
+        # TODO fix expected logits
```
Comment (vasqu) [inline tests/models/jamba/test_modeling_jamba.py:665]: HW-dependent so not sure how to proceed here.

File: tests/models/jamba/test_modeling_jamba.py (hunk 2 above)
Comment (vasqu) [reply tests/models/jamba/test_modeling_jamba.py:505]: Seems like it does :D

**Comment by vasqu** [comment]
CI failure seems unrelated to the PR, some import issues from another model.

File: tests/models/jamba/test_modeling_jamba.py [hunk 4]
```diff
@@ -697,7 +652,7 @@ def test_simple_batched_generate_with_padding(self):
         output_sentences = self.tokenizer.batch_decode(out)
         self.assertEqual(
             output_sentences[0],
-            "<|startoftext|>Hey how are you doing on this lovely evening? Canyon rins hugaughter glamour Rutgers Singh Hebrew cases Cats",
+            "<|startoftext|>Hey how are you doing on this lovely evening? Canyon rins hugaughter glamour Rutgers Singh<|reserved_797|>cw algunas",
```
Comment (vasqu) [inline tests/models/jamba/test_modeling_jamba.py:655]: This might be due to my HW no idea. Would need to see how the slow CI reacts here...

File: src/transformers/models/jamba/modeling_jamba.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/jamba/modeling_jamba.py:1422]: sounds good, let's put the comments in the doc

File: tests/models/jamba/test_modeling_jamba.py (hunk 4 above)
Comment (ArthurZucker) [reply tests/models/jamba/test_modeling_jamba.py:655]: yeah no worries on this one

**Review by ArthurZucker — COMMENTED** [review]
What I would find weird is if this does not improve / change the results. Especially for batched generation! The model is tiny random, would be nice if we can run this with the big one 👀

File: src/transformers/models/jamba/modeling_jamba.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/jamba/modeling_jamba.py:1422]: In which docs exactly? The arguments docs for the attention mask?

**Comment by vasqu** [comment]
Yea, I think it should definitely improve the batched generation. Especially since the `test_left_padding_compatibility` doesn't need higher atols anymore, padding is not as big of a problem as before (I think).

Too GPU poor to run the Jamba models, iirc they require at least an 80GB Vram GPU :cry: Maybe we could notify the guys behind Jamba? I doubt they are aware of this issue.

**Comment by vasqu** [comment]
#32250 seems to have changed the integration tests cc @gante 

Guess we have to redo them again :eyes:

File: tests/models/jamba/test_modeling_jamba.py [hunk 5]
```diff
@@ -737,10 +692,11 @@ def test_simple_batched_generate_with_padding(self):
             with torch.no_grad():
                 logits = self.model(input_ids=inputs["input_ids"]).logits
 
+            # TODO fix logits
```
Comment (vasqu) [inline tests/models/jamba/test_modeling_jamba.py:695]: For more visibility so that I don't forget about it.

File: tests/models/jamba/test_modeling_jamba.py (hunk 2 above)
Comment (vasqu) [reply tests/models/jamba/test_modeling_jamba.py:505]: Keeping it open for visibility: Left padding works fine now, it was an issue of how padding has been handled in general (for mamba-related models).

File: src/transformers/models/jamba/modeling_jamba.py [hunk 6]
```diff
@@ -1383,6 +1410,17 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
 
         return causal_mask
 
+    def _update_mamba_mask(self, attention_mask, cache_position):
+        mamba_mask = attention_mask
+
+        # No need for zeroing states when
+        # 1. Cached forward
+        # 2. Attending to all inputs
+        if cache_position[0] > 0 or (attention_mask is not None and torch.all(attention_mask == 1)):
```
Comment (gante) [inline src/transformers/models/jamba/modeling_jamba.py:1419]: I suspect this line will fail at compilation time (data-dependent conditional branch). Can you confirm, i.e. try running a compiled forward pass?

If it fails, we can add a compile guard, i.e. start the `if` with `not is_torchdynamo_compiling()`

**Review by gante — APPROVED** [review]
LGTM, thank you for fixing! 🙌 

Added a nit to confirm. Pre-approving assuming the logits tests will be addressed (sorry about that :) )

File: src/transformers/models/jamba/modeling_jamba.py (hunk 6 above)
Comment (vasqu) [reply src/transformers/models/jamba/modeling_jamba.py:1419]: Tested via the following mini script:
```python
import torch
from transformers import JambaForCausalLM, AutoTokenizer

model_id = "ai21labs/Jamba-tiny-random"
model = JambaForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, low_cpu_mem_usage=True).to("cuda")
model = torch.compile(model)
tokenizer = AutoTokenizer.from_pretrained(model_id)

# tested on both, batched or non-batched input
#input = tokenizer(["Hey how are you doing on this lovely evening?", "What is the purpose of life?"], padding=True, return_tensors="pt").to("cuda")
input = tokenizer(["What is the purpose of life?"], padding=True, return_tensors="pt").to("cuda")

# tested on both, forward call or generate
out = model(**input)
#out = model.generate(**input, do_sample=False, max_new_tokens=10)
```

Haven't encountered any compilation errors locally, so seems to be fine. Is this what you had in mind to test compilation?

Comment (gante) [reply src/transformers/models/jamba/modeling_jamba.py:1419]: Yes, that's it! 

Perfect, thank you for confirming :)

File: src/transformers/models/jamba/modeling_jamba.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/jamba/modeling_jamba.py:1422]: I mean this:
```python
    def _update_mamba_mask(self, attention_mask, cache_position):
        """
                # No need for zeroing states when
                # 1. Cached forward
                # 2. Attending to all inputs
        """
        mamba_mask = attention_mask
        if cache_position[0] > 0 or (attention_mask is not None and torch.all(attention_mask == 1)):
            mamba_mask = None
        return mamba_mask
```

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks again @vasqu for your great contributions!
